### PR TITLE
Feature/safecdx 771 new

### PR DESCRIPTION
**Pull request by santoshjami**

### CODE REVIEW

File: opencdx-client/src/main/java/cdx/opencdx/client/service/OpenCDXAnalysisEngineClient.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+package cdx.opencdx.client.service;
+
+import cdx.opencdx.client.dto.OpenCDXCallCredentials;
+import cdx.opencdx.client.exceptions.OpenCDXClientException;
+import cdx.opencdx.grpc.service.health.*;
+
+/**
+ * Interface for communicating with the Temperature service.
+ */
+public interface OpenCDXAnalysisEngineClient {
```
Comment (Ghost-Programmer) [inline opencdx-client/src/main/java/cdx/opencdx/client/service/OpenCDXAnalysisEngineClient.java:25]: opencdx-client/src/main/java/cdx/opencdx/client/config/ClientConfig.java  requires changes to support the new client.

File: opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+@GRpcService
+@Observed(name = "opencdx")
+public class OpenCDXAnalysisEngineGrpcController extends AnalysisEngineServiceGrpc.AnalysisEngineServiceImplBase {
+    private final OpenCDXAnalysisEngineService openCDXAnalysisEngineService;
+
+    /**
+     * Constructor using the OpenCDXAnalysisEngineService
+     * @param openCDXAnalysisEngineService service to use for processing
+     */
+    @Autowired
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java:39]: Missing  @Secured({})

File: opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    public OpenCDXAnalysisEngineGrpcController(OpenCDXAnalysisEngineService openCDXAnalysisEngineService) {
+        this.openCDXAnalysisEngineService = openCDXAnalysisEngineService;
+    }
+
+    @Override
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java:44]: Missing  @Secured({})

File: opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    public void createAnalysisEngine(
+            CreateAnalysisEngineRequest request, StreamObserver<CreateAnalysisEngineResponse> responseObserver) {
+        responseObserver.onNext(this.openCDXAnalysisEngineService.createAnalysisEngine(request));
+        responseObserver.onCompleted();
+    }
+
+    @Override
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java:51]: Missing  @Secured({})

File: opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    public void getAnalysisEngine(
+            GetAnalysisEngineRequest request, StreamObserver<GetAnalysisEngineResponse> responseObserver) {
+        responseObserver.onNext(this.openCDXAnalysisEngineService.getAnalysisEngine(request));
+        responseObserver.onCompleted();
+    }
+
+    @Override
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java:58]: Missing  @Secured({})

File: opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    public void updateAnalysisEngine(
+            UpdateAnalysisEngineRequest request, StreamObserver<UpdateAnalysisEngineResponse> responseObserver) {
+        responseObserver.onNext(this.openCDXAnalysisEngineService.updateAnalysisEngine(request));
+        responseObserver.onCompleted();
+    }
+
+    @Override
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java:65]: Missing  @Secured({})

File: opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    public void deleteAnalysisEngine(
+            DeleteAnalysisEngineRequest request, StreamObserver<SuccessResponse> responseObserver) {
+        responseObserver.onNext(this.openCDXAnalysisEngineService.deleteAnalysisEngine(request));
+        responseObserver.onCompleted();
+    }
+
+    @Override
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java:72]: Missing  @Secured({})

File: opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+public class OpenCDXAnalysisEngineModel {
+    @Id
+    private OpenCDXIdentifier id;
+
+    @Version
+    private long version;
+
+    private String name;
+
+    private String organizationId;
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java:49]: This should be an OpenCDXIdentifier not a string.

File: opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    private String workspaceId;
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java:50]: This should be an OpenCDXIdentifier not a string.

File: opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        return builder.build();
+    }
+
+    /**
+     * Updates the OpenCDXAnalysisEngineModel with the given OpenCDXAnalysisEngine object.
+     *
+     * @param analysisEngine The analysisEngine object to update the model with.
+     * @return The updated OpenCDXTemperatureMeasurementModel instance.
+     */
+    public OpenCDXAnalysisEngineModel update(AnalysisEngine analysisEngine) {
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java:137]: This should only update the "name" and "identifier" fields only, the others should not be able to be changed.

File: opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+ * Model for Analysis Engine in Mongo. Features conversions to/from Protobuf messages.
+ */
+@Slf4j
+@Data
+@Builder
+@AllArgsConstructor
+@RequiredArgsConstructor
+@Document("analysis-engine")
+@SuppressWarnings({"java:S3776", "java:S1117", "java:S116"})
+public class OpenCDXAnalysisEngineModel {
```
Comment (Ghost-Programmer) [inline opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java:40]: Missing the identifier field from the proto

**Review by Ghost-Programmer — CHANGES_REQUESTED** [review]

**Comment by Ghost-Programmer** [comment]
Also make sure you check "./deploy.sh"  for a sonarlint issues.

File: opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java (hunk 2 above)
Comment (santoshjami) [reply opencdx-health/src/main/java/cdx/opencdx/health/controller/OpenCDXAnalysisEngineGrpcController.java:39]: Added this attribute on all the grpc controller operations.
@Ghost-Programmer  -  Should this be added on the Rest controller operations as well?

Also this attribute is missing currently in majority of the existing controller classes, may be we need to address them individually.

File: opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java (hunk 9 above)
Comment (santoshjami) [reply opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java:50]: Done

File: opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java (hunk 8 above)
Comment (santoshjami) [reply opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java:49]: Done

File: opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java (hunk 10 above)
Comment (santoshjami) [reply opencdx-health/src/main/java/cdx/opencdx/health/model/OpenCDXAnalysisEngineModel.java:137]: Yes, it is doing the same - updating the name and rest OpenCDXIdentifier fields only (id, orgid, workspaceId)

**Review by Ghost-Programmer — COMMENTED** [review]
What you have here is good.  However we need to discuss this item on the ticket:

"In cdx.opencdx.classification.service.impl.OpenCDXClassificationServiceImpl#classify()  update this method to get the Organization / Workspace from the patient, then do a look up of the AnalysisEngine to use to classify.   This method is doing something similiar cdx.opencdx.health.service.impl.OpenCDXConnectedLabServiceImpl#submitLabFindings"

**Review by Ghost-Programmer — APPROVED** [review]

**Pull request merged**